modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala

```diff
@@ -181,6 +187,69 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
   ): F[Either[ResolutionError, Json]] =
     lookupSchemaResult(schemaKey).map(_.map(_.value.schema))
 
+  /**
+   * If Iglu Central or any of its mirrors doesn't have a schema,
+   * it should be considered NotFound, even if one of them returned an error.
+   */
+  private[resolver] def isNotFound(error: ResolutionError): Boolean = {
+    val (igluCentral, custom) = error.value.partition { case (repo, _) =>
+      allIgluCentral.contains(repo)
+    }
+    (igluCentral.isEmpty || igluCentral.values.exists(
+      _.errors.forall(_ == RegistryError.NotFound)
+    )) && custom.values.flatMap(_.errors).forall(_ == RegistryError.NotFound)
+  }
+
+  /**
+   * Looks up all the schemas with the same model until `maxSchemaKey`.
+   * For the schemas of previous revisions, it starts with addition = 0
+   * and increments it until a NotFound.
+   *
+   * @param maxSchemaKey The SchemaKey until which schemas of the same model should get returned
+   * @return All the schemas if all went well, [[Resolver.SchemaResolutionError]] with the first error that happened
+   *         while looking up the schemas if something went wrong.
+   */
+  def lookupSchemasUntil(
+    maxSchemaKey: SchemaKey
+  )(implicit
+    F: Monad[F],
+    L: RegistryLookup[F],
+    C: Clock[F]
+  ): F[Either[SchemaResolutionError, List[SelfDescribingSchema[Json]]]] = {
+    def go(
+      current: SchemaVer.Full,
+      acc: List[SelfDescribingSchema[Json]]
+    ): F[Either[SchemaResolutionError, List[SelfDescribingSchema[Json]]]] = {
+      val currentSchemaKey = maxSchemaKey.copy(version = current)
+      lookupSchema(currentSchemaKey).flatMap {
+        case Left(e) =>
+          if (current.addition === 0)
+            Monad[F].pure(Left(SchemaResolutionError(currentSchemaKey, e)))
+          else if (current.revision < maxSchemaKey.version.revision && isNotFound(e))
+            go(current.copy(revision = current.revision + 1, addition = 0), acc)
+          else
+            Monad[F].pure(Left(SchemaResolutionError(currentSchemaKey, e)))
+        case Right(json) =>
+          if (current.revision < maxSchemaKey.version.revision)
+            go(
+              current.copy(addition = current.addition + 1),
+              SelfDescribingSchema(SchemaMap(currentSchemaKey), json) :: acc
+            )
+          else if (current.addition < maxSchemaKey.version.addition)
+            go(
+              current.copy(addition = current.addition + 1),
+              SelfDescribingSchema(SchemaMap(currentSchemaKey), json) :: acc
+            )
+          else
+            Monad[F].pure(
+              Right((SelfDescribingSchema(SchemaMap(currentSchemaKey), json) :: acc).reverse)
+            )
+      }
+    }
+
+    go(SchemaVer.Full(maxSchemaKey.version.model, 0, 0), Nil)
```

Comment:
Hi @istreeter ,
I now want to return a `NonEmptyVector`. I see two approaches:
1. Manually fetch `model-0-0` and then call `go()` with a `NonEmptyVector` as `acc`
2. Keep the existing and map the result like
```
  case head :: tail => create non empty vector
  case _ => create SchemaResolutionError, although we know that it's not possible
```
I don't like both approaches, but I'd still pick 2 as I don't like to need some action before calling `go()`. WDYT ?

Reply:
Sounds like 1 is the correct way to do it.  Unless that becomes horrifically ugly.  But I'm hopeful there will be a way to do 1 without ruining the simplicity of the code.